src/main.c

```diff
@@ -142,10 +147,160 @@ static bool is_valid_stream(char *s)
 	return true;
 }
 
+char* get_saved_file_path(void)
+{
+	const char* xdg_state_home = getenv("XDG_STATE_HOME");
+	const char* home_path = getenv("HOME");
+
+	if (!home_path & !xdg_state_home)
+	exit(EXIT_FAILURE);
+
+	const char* local_state = "/.local/state";
+	const char* dir_name = "/nudoku/";
+
+	char* dir_path = NULL;
+
+	if (xdg_state_home)
+	{
+	// user does have $XDG_STATE_HOME
+	size_t len_xdg_state_home = strlen(xdg_state_home);
+	size_t len_dir_name = strlen(dir_name);
+	dir_path = malloc(len_xdg_state_home + len_dir_name + 1);
+
+	strcpy(dir_path, xdg_state_home);
+	strcat(dir_path, dir_name);
+	}
+	else
+	{
+	// fallback to $HOME/.local/state
+	size_t len_home_path = strlen(home_path);
+	size_t len_local_state = strlen(local_state);
+	size_t len_dir_name = strlen(dir_name);
+	dir_path = malloc(len_home_path + len_local_state + len_dir_name + 1);
+
+	strcpy(dir_path, home_path);
+	strcat(dir_path, local_state);
+	strcat(dir_path, dir_name);
+	}
+
+	struct stat st = {0};
+	if (stat(dir_path, &st) == -1)
+	{
+	mkdir(dir_path, 0700);
+	}
+
+	const char* file_name = STATE_FILE_NAME;
+	size_t len_file = strlen(file_name);
+	size_t len_dir_path = strlen(dir_path);
+
+	char* file_path = malloc(len_dir_path + len_file + 1);
+	strcpy(file_path, dir_path);
+	strcat(file_path, file_name);
+
+	free(dir_path);
+	return file_path;
+}
+
+bool is_saved_game_valid(void)
+{
+	char* file_path = get_saved_file_path();
+	FILE* fp = fopen(file_path, "r");
+
+	if (fseek(fp, STREAM_LENGTH, SEEK_SET) != 0)
+	{
+	fclose(fp);
+	free(file_path);
+	return false;
+	}
+
+	char test_board[STREAM_LENGTH];
+	int c;
+	int i = 0;
+
+	while (i < STREAM_LENGTH)
+	{
+	c = fgetc(fp);
+	if (c == EOF) break;
+	test_board[i++] = (char)c;
+	}
+
+	bool valid = is_valid_stream(test_board);
+
+	fclose(fp);
+	free(file_path);
+
+	return valid;
+}
+
+void save_stream(char user_board[], char plain_board[], int n)
+{
+	char* file_path = get_saved_file_path();
+	FILE* fp;
+	fp = fopen(file_path, "w");
+	if ( fp == NULL )
+		exit(EXIT_FAILURE);
+	for ( int i=0; i<n; ++i )
+	{
+		fprintf(fp, "%c", user_board[i]);
+	}
+	for ( int i=0; i<n; ++i )
+	{
+		fprintf(fp, "%c", plain_board[i]);
+	}
+	fclose(fp);
+	free(file_path);
+}
+
+void get_saved_game_state(char user_board[], char plain_board[])
+{
+	FILE* fp;
```

Comment:
The loading and processing of the saved file is now done twice one time here and one time in `is_saved_game_valid()`.
I would recommend to use a  helper function that reads in the file and stores it in `char board [STREAM_LENGTH]` that the helper function returns.

So that we dont duplicate that code.

We could even think about doing that once, and then passing that board to both the validation function and to this one for reading only once from disk. But I didn't check the code in detail how it will affect it.

But in any case the code duplication should go :)